service_manager/service_manager.py

```diff
@@ -0,0 +1,43 @@
+import asyncio
+from yapapi.log import enable_default_logger
+from .service_wrapper import ServiceWrapper
+from .yapapi_connector import YapapiConnector
+from functools import partial
+
+from typing import TYPE_CHECKING
+if TYPE_CHECKING:
+    from typing import Type, List, Any, Callable, Awaitable
+    from yapapi.services import Service
+
+
+async def stop_on_golem_exception(service_manager: 'ServiceManager', e: Exception) -> None:
+    print("GOLEM FAILED\n", e)
+    print("STOPPING THE LOOP")
+    loop = asyncio.get_event_loop()
+    loop.stop()
+
+
+class ServiceManager():
+    def __init__(
+        self,
+        executor_cfg: dict,
+        golem_exception_handler: 'Callable[[ServiceManager, Exception], Awaitable[None]]' = stop_on_golem_exception,
+        log_file: str = 'log.log'
+    ):
+        enable_default_logger(log_file=log_file)
+
+        self.service_wrappers: 'List[ServiceWrapper]' = []
+        exception_handler = partial(golem_exception_handler, self)
+        self.yapapi_connector = YapapiConnector(executor_cfg, exception_handler)
+
+    def create_service(self, service_cls: 'Type[Service]', start_args: 'List[Any]' = [],
+                       service_wrapper_cls: 'Type[ServiceWrapper]' = ServiceWrapper):
+        service_wrapper = service_wrapper_cls(service_cls, start_args)
```

Comment:
What is the reason to instantiate `service_wrapper` here instead of simply accepting it as a parameter?

Reply:
1. The end-user interface has one import less and one object initialization less
2. It's harder to make an error by creating `ServiceWrapper()` object but **not** passing it to the `service_manager`.
3. We might want to change the signature of `ServiceWrapper`. E.g. `id` could be passed by `service_manager` (from some pool, e.g. `service_cls`-dependent). As long as we have this sort of a wrapper function, it would be much easier to maintain backward compatibility (yes, I remember we don't guarantee this, but even for our own uses like `Erigon` or `yagna-requests` it will be a nice feature).
4. We might consider something like this:
```
def create_service(self, some_args):
     similar_service_wrapper = self.find_matching_service_wrapper(some_args)
     if similar_service_wrapper is not None:
         similar_service_wrapper.cluster.add_new_instance()
         return similar_service_wrapper
     else:
         new_service_wrapper = ...
         return new_service_wrapper
```
(I'm not saying this is a good idea now, but we might want to go in this direction)

On the other hand, the serious downside is that user can't pass any additional args to `__init__`.
I would leave this function as it is (for the sake of p.1 & 2 above), but there might be another path in the future, e.g.
```
service_wrapper = ServiceWrapper(...)
service_wrapper.create_new_cluster(service_manager)
```

Reply:
> 1. The end-user interface has one import less and one object initialization less

One initialization, yes. But import? I don't see which import is spared.

> 2. It's harder to make an error by creating ServiceWrapper() object but not passing it to the service_manager.

That's true but I wouldn't take it as a serious issue.

> 3. We might want to change the initialization of `ServiceWrapper`.

Making constructor a part of the interface is not a good idea. If we want to go in this direction I recommend changing this parameter from class into 'factory method'.

Reply:
4132fc2